src/quiz/use-quiz.ts

```diff
@@ -1,23 +1,70 @@
-import { useState } from "react";
+import { useEffect, useState } from "react";
 
 import { type Question } from "./types";
 
-export const useQuiz = (initialQuestions: Question[]) => {
+interface Props {
+	initialQuestions: Question[];
+	validationMessages: {
+		correct: string;
+		incorrect: string;
+	};
+}
+
+export const useQuiz = ({ initialQuestions, validationMessages }: Props) => {
 	const [quizAnswers, setQuizAnswers] = useState<(number | undefined)[]>(
 		initialQuestions.map((question) => question.selectedAnswer),
 	);
+	const [questions, setQuestions] = useState<Question[]>([]);
+	const [correctAnswerCount, setCorrectAnswerCount] = useState(0);
+
+	// Initialize the `questions` state and make it synchronized with `quizAnswers`.
+	// The synchronization is needed in order to reflect the correct `selectedAnswer`.
+	useEffect(() => {
+		const questionsWithChangeHandling = initialQuestions.map(
+			(question, index) => ({
+				...question,
+				onChange: (selectedAnswer: number) => {
+					setQuizAnswers((prevAnswers) =>
+						prevAnswers.map((prevAnswer, prevIndex) =>
+							prevIndex === index ? selectedAnswer : prevAnswer,
+						),
+					);
+				},
+				selectedAnswer: quizAnswers[index],
+			}),
+		);
+
+		setQuestions(questionsWithChangeHandling);
+		// We only need `quizAnswers` in the dependency array.
+		// Adding `initialQuestions` in will cause an infinite loop.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [quizAnswers]);
+
+	const validateAnswers = () => {
+		setQuestions((prevQuestions) =>
+			prevQuestions.map((prevQuestion) => {
+				if (prevQuestion.selectedAnswer === prevQuestion.correctAnswer) {
+					setCorrectAnswerCount((count) => count + 1);
+
+					return {
+						...prevQuestion,
+						validation: {
+							state: "correct",
+							message: validationMessages.correct,
+						},
+					};
+				}
+
+				return {
+					...prevQuestion,
+					validation: {
+						state: "incorrect",
+						message: validationMessages.incorrect,
+					},
+				};
+			}),
+		);
```

Comment:
```suggestion
		setQuestions((prevQuestion) => {
			const updatedQuestions: Question[] = prevQuestion.map((question) => {
				const validation: Question["validation"] =
					question.selectedAnswer === question.correctAnswer
						? {
								state: "correct",
								message: validationMessages.correct,
							}
						: {
								state: "incorrect",
								message: validationMessages.incorrect,
							};
				return { ...question, validation };
			});

			const correctCount = updatedQuestions.filter(
				({ validation }) => validation?.state === "correct",
			).length;
			setCorrectAnswerCount(correctCount);
			return updatedQuestions;
		});
```
There's not much difference, but I like the two step (update validation, calculate new count) better than doing it all at once. Up to you, though.